dart/common/detail/Addon.h

```diff
@@ -0,0 +1,338 @@
+{
+  DART_COMMON_CAST_NEW_MANAGER_TYPE_AND_RETURN_NULL_IF_BAD(
+        Base, ManagerType, newManager, castedManager, clone);
+  return std::unique_ptr<Base>(new Base(newManager, mState));
+}
+
+//==============================================================================
+template <class Base, typename PropertiesData,
+          class ManagerType, void (*updateProperties)(Base*)>
+AddonWithProtectedProperties<Base, PropertiesData,
+                             ManagerType, updateProperties>::
+AddonWithProtectedProperties(
+    AddonManager* mgr, const PropertiesData& properties)
+  : Addon(mgr),
+    mProperties(properties)
+{
+  // Do nothing
+}
+
+//==============================================================================
+template <class Base, typename PropertiesData,
+          class ManagerType, void (*updateProperties)(Base*)>
+void AddonWithProtectedProperties<Base, PropertiesData,
+                                  ManagerType, updateProperties>::
+setAddonProperties(const std::unique_ptr<Addon::Properties>& someProperties)
+{
+  setProperties(static_cast<const Properties&>(*someProperties));
+}
+
+//==============================================================================
+template <class Base, typename PropertiesData,
+          class ManagerType, void (*updateProperties)(Base*)>
+const Addon::Properties*
+AddonWithProtectedProperties<Base, PropertiesData,
+                             ManagerType, updateProperties>::
+getAddonProperties() const
+{
+  return &mProperties;
+}
+
+//==============================================================================
+template <class Base, typename PropertiesData,
+          class ManagerType, void (*updateProperties)(Base*)>
+void AddonWithProtectedProperties<Base, PropertiesData,
+                                  ManagerType, updateProperties>::
+setProperties(const PropertiesData& properties)
+{
+  static_cast<PropertiesData&>(mProperties) = properties;
+  UpdateProperties(static_cast<Base*>(this));
+}
+
+//==============================================================================
+template <class Base, typename PropertiesData,
+          class ManagerType, void (*updateProperties)(Base*)>
+auto AddonWithProtectedProperties<Base, PropertiesData,
+                                  ManagerType, updateProperties>::
+getProperties() const -> const Properties&
+{
+  return mProperties;
+}
+
+//==============================================================================
+template <class Base, typename PropertiesData,
+          class ManagerType, void (*updateProperties)(Base*)>
+std::unique_ptr<Addon>
+AddonWithProtectedProperties<Base, PropertiesData,
+                             ManagerType, updateProperties>::
+cloneAddon(AddonManager* newManager) const
+{
+  DART_COMMON_CAST_NEW_MANAGER_TYPE_AND_RETURN_NULL_IF_BAD(
+        Base, ManagerType, newManager, castedManager, clone);
+  return std::unique_ptr<Base>(new Base(newManager, mProperties));
```

Comment:
This is potentially dangers. You should [*always* parenthesis macro arguments that could be an expression](http://stackoverflow.com/a/23889851/111426):
```c++
return std::unique_ptr<Base>(new Base((newManager), mProperties));
```